File: expense-tracker/routes/accounts.py
from flask import Blueprint, render_template, request, redirect, url_for, session, flash
from database.db import get_db
from helpers import CURRENCY_CHOICES, currency_symbol

accounts_bp = Blueprint('accounts', __name__)

ACCOUNT_TYPES = ['bank', 'cash', 'credit_card', 'investment', 'wallet', 'other']


@accounts_bp.route('/accounts', endpoint='list_accounts')
def list_accounts():
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    user_id = session['user_id']
    accounts = db.execute(
        "SELECT * FROM accounts WHERE user_id = ? ORDER BY is_active DESC, name ASC",
        (user_id,)
    ).fetchall()

    # Get balances from actual transactions
    for acc in accounts:
        spent = db.execute(
            "SELECT COALESCE(SUM(amount), 0) as total FROM expenses WHERE user_id = ? AND account_id = ?",
            (user_id, acc['id'])
        ).fetchone()['total'] or 0.0
        earned = db.execute(
            "SELECT COALESCE(SUM(amount), 0) as total FROM income WHERE user_id = ? AND account_id = ?",
            (user_id, acc['id'])
        ).fetchone()['total'] or 0.0
        acc['calculated_balance'] = earned - spent

    user = db.execute("SELECT preferred_currency FROM users WHERE id = ?", (user_id,)).fetchone()
    preferred_currency = user['preferred_currency'] if user and user['preferred_currency'] else 'INR'

    return render_template('accounts.html', accounts=accounts, account_types=ACCOUNT_TYPES,
                           currencies=CURRENCY_CHOICES, preferred_currency=preferred_currency,
                           currency_symbol=currency_symbol)


@accounts_bp.route('/accounts/add', methods=['POST'], endpoint='add_account')
def add_account():
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    name = (request.form.get('name') or '').strip()
    acc_type = request.form.get('type', 'bank')
    currency = request.form.get('currency', 'INR')

    if not name or len(name) < 1:
        flash('Account name is required.', 'danger')
        return redirect(url_for('accounts.list_accounts'))
    if acc_type not in ACCOUNT_TYPES:
        acc_type = 'bank'

    db = get_db()
    db.execute(
        "INSERT INTO accounts (user_id, name, type, currency) VALUES (?, ?, ?, ?)",
        (session['user_id'], name, acc_type, currency)
    )
    db.commit()
    flash(f'Account "{name}" created!', 'success')
    return redirect(url_for('accounts.list_accounts'))


@accounts_bp.route('/accounts/<int:id>/edit', methods=['POST'], endpoint='edit_account')
def edit_account(id):
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    account = db.execute("SELECT * FROM accounts WHERE id = ? AND user_id = ?", (id, session['user_id'])).fetchone()
    if not account:
        flash('Account not found.', 'danger')
        return redirect(url_for('accounts.list_accounts'))

    name = (request.form.get('name') or '').strip()
    acc_type = request.form.get('type', 'bank')
    currency = request.form.get('currency', 'INR')
    is_active = 1 if request.form.get('is_active') else 0

    if not name:
        flash('Account name is required.', 'danger')
        return redirect(url_for('accounts.list_accounts'))

    db.execute(
        "UPDATE accounts SET name = ?, type = ?, currency = ?, is_active = ? WHERE id = ?",
        (name, acc_type, currency, is_active, id)
    )
    db.commit()
    flash('Account updated!', 'success')
    return redirect(url_for('accounts.list_accounts'))


@accounts_bp.route('/accounts/<int:id>/delete', methods=['POST'], endpoint='delete_account')
def delete_account(id):
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    account = db.execute("SELECT * FROM accounts WHERE id = ? AND user_id = ?", (id, session['user_id'])).fetchone()
    if not account:
        flash('Account not found.', 'danger')
        return redirect(url_for('accounts.list_accounts'))

    # Unlink transactions from this account
    db.execute("UPDATE expenses SET account_id = NULL WHERE account_id = ? AND user_id = ?", (id, session['user_id']))
    db.execute("UPDATE income SET account_id = NULL WHERE account_id = ? AND user_id = ?", (id, session['user_id']))
    db.execute("DELETE FROM accounts WHERE id = ?", (id,))
    db.commit()
    flash(f'Account "{account["name"]}" deleted.', 'info')
    return redirect(url_for('accounts.list_accounts'))

File: backend/routes/accounts.js
const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');
const { Account, Expense, Income, User } = require('../models');
const { CURRENCY_CHOICES } = require('../src/helpers');

const ACCOUNT_TYPES = ['bank', 'cash', 'credit_card', 'investment', 'wallet', 'other'];

router.get('/accounts', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;

    try {
        const accounts = await Account.find({ user_id }).sort({ is_active: -1, name: 1 });
        
        const spentByAccount = await Expense.aggregate([
            { $match: { user_id: new mongoose.Types.ObjectId(user_id), account_id: { $ne: null } } },
            { $group: { _id: '$account_id', total: { $sum: '$amount' } } }
        ]);

        const earnedByAccount = await Income.aggregate([
            { $match: { user_id: new mongoose.Types.ObjectId(user_id), account_id: { $ne: null } } },
            { $group: { _id: '$account_id', total: { $sum: '$amount' } } }
        ]);

        const spentMap = {};
        for (const row of spentByAccount) {
            if (row._id) spentMap[row._id.toString()] = row.total;
        }

        const earnedMap = {};
        for (const row of earnedByAccount) {
            if (row._id) earnedMap[row._id.toString()] = row.total;
        }

        const accounts_data = accounts.map(acc => {
            const accIdStr = acc._id.toString();
            const spent = spentMap[accIdStr] || 0;
            const earned = earnedMap[accIdStr] || 0;
            const accObj = acc.toObject();
            accObj.id = accIdStr;
            accObj.calculated_balance = earned - spent;
            return accObj;
        });

        const user = await User.findById(user_id);
        const preferred_currency = user?.preferred_currency || 'INR';

        res.render('accounts.html', {
            accounts: accounts_data,
            account_types: ACCOUNT_TYPES,
            currencies: CURRENCY_CHOICES,
            preferred_currency
        });
    } catch (err) {
        next(err);
    }
});

router.post('/accounts/add', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const name = (req.body.name || '').trim();
    let acc_type = req.body.type || 'bank';
    const currency = req.body.currency || 'INR';

    if (!name || name.length < 1) {
        req.flash('danger', 'Account name is required.');
        return res.redirect('/accounts');
    }
    if (!ACCOUNT_TYPES.includes(acc_type)) {
        acc_type = 'bank';
    }

    try {
        const newAcc = new Account({
            user_id,
            name,
            type: acc_type,
            currency
        });
        await newAcc.save();
        req.flash('success', `Account "${name}" created!`);
        res.redirect('/accounts');
    } catch (err) {
        next(err);
    }
});

router.post('/accounts/:id/edit', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const account_id = req.params.id;

    if (!mongoose.isValidObjectId(account_id)) {
        req.flash('danger', 'Account not found.');
        return res.redirect('/accounts');
    }

    const name = (req.body.name || '').trim();
    let acc_type = req.body.type || 'bank';
    const currency = req.body.currency || 'INR';
    const is_active = !!req.body.is_active;

    if (!name) {
        req.flash('danger', 'Account name is required.');
        return res.redirect('/accounts');
    }
    if (!ACCOUNT_TYPES.includes(acc_type)) {
        acc_type = 'bank';
    }

    try {
        const account = await Account.findOne({ _id: account_id, user_id });
        if (!account) {
            req.flash('danger', 'Account not found.');
            return res.redirect('/accounts');
        }

        account.name = name;
        account.type = acc_type;
        account.currency = currency;
        account.is_active = is_active;

        await account.save();
        req.flash('success', 'Account updated!');
        res.redirect('/accounts');
    } catch (err) {
        next(err);
    }
});

router.post('/accounts/:id/delete', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const account_id = req.params.id;

    if (!mongoose.isValidObjectId(account_id)) {
        req.flash('danger', 'Account not found.');
        return res.redirect('/accounts');
    }

    try {
        const account = await Account.findOne({ _id: account_id, user_id });
        if (!account) {
            req.flash('danger', 'Account not found.');
            return res.redirect('/accounts');
        }

        // Unlink transactions
        await Expense.updateMany({ account_id, user_id }, { account_id: null });
        await Income.updateMany({ account_id, user_id }, { account_id: null });

        await Account.deleteOne({ _id: account_id, user_id });
        req.flash('info', `Account "${account.name}" deleted.`);
        res.redirect('/accounts');
    } catch (err) {
        next(err);
    }
});

// ------------------------------------------------------------------ //
// JSON API Endpoint                                                  //
// ------------------------------------------------------------------ //
router.get('/api/accounts', async (req, res) => {
    if (!req.session.user_id) {
        return res.status(401).json({ error: 'Not authenticated' });
    }
    try {
        const accounts = await Account.find({ user_id: req.session.user_id }).sort({ is_active: -1, name: 1 }).lean();
        const spentByAccount = await Expense.aggregate([
            { $match: { user_id: new mongoose.Types.ObjectId(req.session.user_id), account_id: { $ne: null } } },
            { $group: { _id: '$account_id', total: { $sum: '$amount' } } }
        ]);
        const earnedByAccount = await Income.aggregate([
            { $match: { user_id: new mongoose.Types.ObjectId(req.session.user_id), account_id: { $ne: null } } },
            { $group: { _id: '$account_id', total: { $sum: '$amount' } } }
        ]);
        const spentMap = {};
        for (const row of spentByAccount) if (row._id) spentMap[row._id.toString()] = row.total;
        const earnedMap = {};
        for (const row of earnedByAccount) if (row._id) earnedMap[row._id.toString()] = row.total;

        const accounts_data = accounts.map(acc => ({
            id: acc._id.toString(),
            name: acc.name,
            type: acc.type,
            currency: acc.currency || 'INR',
            is_active: acc.is_active,
            calculated_balance: (earnedMap[acc._id.toString()] || 0) - (spentMap[acc._id.toString()] || 0)
        }));
        res.json(accounts_data);
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

module.exports = router;
